lib/puppet/type.rb

```diff
@@ -2117,8 +2117,17 @@ def autorelation(rel_type, rel_catalog = nil)
     auto_rel = "eachauto#{rel_type}".to_sym
 
     self.class.send(auto_rel) { |type, block|
-      # Ignore any types we can't find, although that would be a bit odd.
-      next unless Puppet::Type.type(type)
+      if not Puppet::Type.type(type)
+        # We know it is not a type at this point, let's see if it exists as a defined type in the catalog
+        # We split and join on :: because defined types will likely have colon separators
+        # All defined types are of type :component
+        found_defines = rel_catalog.resources.select do |x|
+          x.type == :component && x.title_class == type.to_s.split('::').map(&:capitalize).join('::')
```

Comment:
I still have to write a Jira feature request for this in full description but see the note on the commit for some explanation.  One spot is that mighq/ipset module uses a defined type instead of a custom type.  IPset is an extension to iptables and I use it in my firewalld module.  I prefer to be able to grab their ipset name that they're using in their direct rule and do an autorequire on it inside my firewalld custom type to ensure ordering and also have the option for enforcing a hard requires if chosen by utilizing fail.  As the commit note points out.  This was only possible using autorequire(:component) do ... end  before.  This PR would fix that so I could do autorequire(:ipset) (which is a defined type)